deepmd/pt/model/model/pair_tab.py

```diff
@@ -0,0 +1,374 @@
+
+        ----------------------------------------------
+
+        table = [[0.005 1.    2.    3.   ]
+                [0.01  0.8   1.6   2.4  ]
+                [0.015 0.5   1.    1.5  ]
+                [0.02  0.25  0.4   0.75 ]
+                [0.025 0.    0.1   0.   ]
+                [0.03  0.    0.    0.   ]]
+
+        rcut = 0.031
+
+        new_table = [[0.005 1.    2.    3.   ]
+                    [0.01  0.8   1.6   2.4  ]
+                    [0.015 0.5   1.    1.5  ]
+                    [0.02  0.25  0.4   0.75 ]
+                    [0.025 0.    0.1   0.   ]
+                    [0.03  0.    0.    0.   ]
+                    [0.035 0.    0.    0.   ]]
+        """
+        raw_data = np.loadtxt(self.tab_file)
+        upper = raw_data[-1][0]
+        upper_val = raw_data[-1][1:]
+        upper_idx = raw_data.shape[0] - 1
+        increment = raw_data[1][0] - raw_data[0][0]
+
+        # the index of table for the grid point right after rcut
+        rcut_idx = int(self.rcut / increment)
+
+        if np.all(upper_val == 0):
+            # if table values decay to `0` after rcut
+            if self.rcut < upper and np.any(raw_data[rcut_idx - 1][1:] != 0):
+                logging.warning(
+                    "The energy provided in the table does not decay to 0 at rcut."
+                )
+            # if table values decay to `0` at rcut, do nothing
+
+            # if table values decay to `0` before rcut, pad table with `0`s.
+            elif self.rcut > upper:
+                pad_zero = np.zeros((rcut_idx - upper_idx, 4))
+                pad_zero[:, 0] = np.linspace(
+                    upper + increment, increment * (rcut_idx + 1), rcut_idx - upper_idx
+                )
+                raw_data = np.concatenate((raw_data, pad_zero), axis=0)
+        else:
+            # if table values do not decay to `0` at rcut
+            if self.rcut <= upper:
+                logging.warning(
+                    "The energy provided in the table does not decay to 0 at rcut."
+                )
+            # if rcut goes beyond table upper bond, need extrapolation.
+            else:
+                logging.warning(
+                    "The rcut goes beyond table upper boundary, performing linear extrapolation."
+                )
+                pad_linear = np.zeros((rcut_idx - upper_idx + 1, 4))
+                pad_linear[:, 0] = np.linspace(
+                    upper, increment * (rcut_idx + 1), rcut_idx - upper_idx + 1
+                )
+                pad_linear[:, 1:] = np.array(
+                    [
+                        np.linspace(start, 0, rcut_idx - upper_idx + 1)
+                        for start in upper_val
+                    ]
+                ).T
+                raw_data = np.concatenate((raw_data[:-1, :], pad_linear), axis=0)
+
+        # over writing file with padding if applicable.
+        with open(self.tab_file, "wb") as f:
+            np.savetxt(f, raw_data)
```

Comment:
Overwriting the user's provided file is not a good idea.

Reply:
three options here:

1. overwriting the file: I agree, this may not be the best practice, even though all the original information is still kept in the file.
2. write to a new file
3. change the `PairTab` and allow both `file path` and `np.array` as input. With this, we can easily add test cases, without writing new files.

Reply:
@njzjz @wanghan-iapcm I lean towards option 3

Reply:
It should not be the input of `PairTab`. Instead, it should be internal within `PairTab`.